src/guest_memory.rs

```diff
@@ -646,10 +646,15 @@ pub trait GuestMemory {
                 Ok(0) => return Ok(total),
                 // made some progress
                 Ok(len) => {
-                    total += len;
-                    if total == count {
-                        break;
-                    }
+                    match total.checked_add(len) {
+                        Some(x) if x <= count => {
+                            total = x;
+                            if total == count {
+                                break;
+                            }
+                        },
+                        _ => return Err(Error::InvalidBackendAddress),
+                    };
                     cur = match cur.overflowing_add(len as GuestUsize) {
```

Comment:
Hm... Looking at it, it _looks_ like it would panic before or immediately after the above code would panic. Should we maybe write a test for this? We can mock fake regions, right?

Reply:
Ideally it shouldn't ever panic. It should always try to return an error that can be properly handled and propagated.

It should not panic below.

Reply:
> Ideally it shouldn't ever panic. It should always try to return an error that can be properly handled and propagated.
> 
> It should not panic below.

Full ACK. But that's why I think both additions / checks should probably be touched. And I think a test would help a lot here.

Reply:
I've updated the PR to avoid the panic.